R/ggplotly.R

```diff
@@ -459,27 +459,35 @@ gg2list <- function(p, width = NULL, height = NULL,
   
   # panel -> plotly.js axis/anchor info
   # (assume a grid layout by default)
-  layout$layout$xaxis <- layout$layout$COL
-  layout$layout$yaxis <- layout$layout$ROW
-  layout$layout$xanchor <- nRows
-  layout$layout$yanchor <- 1
+  layout$layout <- dplyr::mutate(layout$layout,
+    xaxis = COL,
+    yaxis = ROW,
+    xanchor = nRows,
+    yanchor = 1L)
   if (inherits(plot$facet, "FacetWrap")) {
-    if (plot$facet$params$free$x) {
-      layout$layout$xaxis <- layout$layout$PANEL
-      layout$layout$xanchor <- layout$layout$ROW
-    }
-    if (plot$facet$params$free$y) {
-      layout$layout$yaxis <- layout$layout$PANEL
-      layout$layout$yanchor <- layout$layout$COL
-      layout$layout$xanchor <- nPanels
-    }
     if (plot$facet$params$free$x && plot$facet$params$free$y) {
-      layout$layout$xaxis <- layout$layout$PANEL
-      layout$layout$yaxis <- layout$layout$PANEL
-      layout$layout$xanchor <- layout$layout$PANEL
-      layout$layout$yanchor <- layout$layout$PANEL
+      layout$layout <- dplyr::mutate(layout$layout,
+        xaxis = PANEL,
+        yaxis = PANEL,
+        xanchor = PANEL,
+        yanchor = PANEL)
+    } else if (plot$facet$params$free$x) {
+      layout$layout <- dplyr::mutate(layout$layout,
+        xaxis = PANEL,
+        xanchor = ROW)
+    } else if (plot$facet$params$free$y) {
+      layout$layout <- dplyr::mutate(layout$layout,
+        yaxis = PANEL,
+        yanchor = COL)
     }
+    # anchor X axis to the lowest plot in its column
+    layout$layout <- dplyr::group_by(layout$layout, xaxis) %>%
+      dplyr::mutate(xanchor = max(as.integer(yaxis))) %>%
+      dplyr::ungroup() %>%
+      dplyr::mutate(xanchor = if (is.factor(yaxis)) levels(yaxis)[xanchor] else xanchor)
   }
```

Comment:
I don't immediately follow why `dplyr::mutate(xanchor = if (is.factor(yaxis)) levels(yaxis)[xanchor] else xanchor)` is necessary, do you have an example in mind where this becomes relevant?

Reply:
Yes, it looks a bit ugly. I had trouble with some of the plots (maybe even the ones in the tests) when either a PANEL or COL is a factor.
There was some error happening afterwards, but unfortunately I forgot the details. So I had to convert `xanchor` back to a factor.

Reply:
@alyst tests are passing without it -- I would be very grateful if you can report a case where 3c8505230aca09c4adf4672f91ffc1e1cc4a32f0 doesn't do what you'd expect!